modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/dataRenderers.ts

```diff
@@ -0,0 +1,131 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import {FDSCellRenderer} from '@liferay/js-api/data-set';
+import React from 'react';
+
+// @ts-ignore
+
+import {dataRenderers, inputRenderers} from '../data_renderers/index';
+
+// @ts-ignore
+
+import {getJsModule} from './modules';
+
+export function getInputRendererById(id: string): any {
+	const inputRenderer = inputRenderers[id];
+
+	if (!inputRenderer) {
+		throw new Error(`No input renderer found with id "${id}"`);
+	}
+
+	return inputRenderer;
+}
+
+export type DataRendererType = 'clientExtension' | 'internal';
+
+export interface DataRenderer {
+	type: DataRendererType;
+}
+
+export interface InternalDataRenderer extends DataRenderer {
+	Component: React.ComponentClass<any>;
+	type: 'internal';
+}
+
+export interface ClientExtensionDataRenderer extends DataRenderer {
+	renderer: FDSCellRenderer;
+	type: 'clientExtension';
+}
+
+export type AnyDataRenderer =
+	| ClientExtensionDataRenderer
+	| InternalDataRenderer;
```

Comment:
Formatter adds the `|` before the first type, yes. It's exactly the same as `ClientExtensionDataRenderer| InternalDataRenderer` but formatted.

Reply:
>  It's a bit inconsistent naming here that DataRenderer is an interface but AnyDataRenderer a type.

:thinking: An interface is a type. And the `AnyDataRenderer` is what it say "Any" of the available DataRenderers, which could be ClientExtensionDataRenderer or InternalDataRenderer, which are interfaces (thus types) too.

The only difference that makes sense IMO is class vs abstrac type, since you can instantiate a class but not an abstract type...

Reply:
> We also have a dataRenderer that is a component

Nope, `dataRenderer` is a function, that's why it is lowercase. In the InternalDataRenderer case it is a real React component, and we store the class, not the instance, and that's why I used uppercase for that.

Reply:
> I started to apply naming convention like in Clay, with T and I prefixes.

I really dislike this :sweat_smile: People used to do it long ago, but IINM it's less and less used because it doesn't add very much information, makes refactors harder and, if you have a decent compiler, they are redundant. In any case, as always, there are people that like and other than don't so I'll stick to anything we decide :shrug: 

For reference: https://en.wikipedia.org/wiki/Hungarian_notation

Reply:
I completely agree that having Hungarian notation as a comprehensive convention is not necessary. And, i am not a fan of it either.

 But this is a case where I feel we should have _some_ convention. It comes from the "meta" nature of TS, where we have two named descriptors. Maybe the way to go so have `*Props` suffix on all interfaces and `*Type` suffix on all types :thinking: